test/indexer/test_indexer_controller.py

```diff
@@ -0,0 +1,152 @@
+
+    @mock_sts
+    def test_contribution_action_reindex(self):
+        event = [self.generate_sqs_record(action='reindex', prefix='ff')]
+        with mock.patch('azul.azulclient.AzulClient.do_remote_reindex') as mock_reindex:
+            mock_reindex.return_value = True
+            self.controller.contribute(event)
+            self.assertTrue(mock_reindex.called)
+            self.assertEqual(mock_reindex.call_count, 1)
+
+    @mock_sts
+    @mock_sqs
+    def test_contribute_and_aggregate(self):
+        """
+        Index two bundles that make contributions to the same project.
+        Inspect that the contributions match the tallies that are returned to
+        SQS. During aggregation only the project entity is deferred due to
+        multiple contributions.
+        """
+        self._create_mock_queues()
+        events = []
+        bundles = []
+        contributions_per_bundle: Sequence[Sequence[Contribution]] = []
+        bundle_fqids = [
+            BundleFQID('56a338fe-7554-4b5d-96a2-7df127a7640b', '2018-03-28T151023.074974Z'),
+            BundleFQID('b2216048-7eaa-45f4-8077-5a3fb4204953', '2018-03-29T104041.822717Z')
+        ]
+        for bundle_fqid in bundle_fqids:
+            notification = self.client.synthesize_notification(self.catalog, bundle_fqid)
+            events.append(self.generate_sqs_record(action='add',
+                                                   catalog=self.catalog,
+                                                   notification=notification))
+            bundle = self._load_canned_bundle(bundle_fqid)
+            bundles.append(bundle)
+            contributions_per_bundle.append(self.index_service.transform(self.catalog, bundle=bundle, delete=False))
+        with mock.patch('azul.indexer.index_controller.IndexController.repository_plugin') as mock_repository, (
+            mock.patch('azul.indexer.index_controller.IndexController.index_service',
+                       new_callable=mock.PropertyMock)) as mock_index_service:
+            mock_index_service.return_value = ForcedRefreshIndexService()
+            mock_repository.return_value = mock.MagicMock()
+            mock_repository.return_value.fetch_bundle.return_value = yield contributions_per_bundle
+            self.controller.contribute(events)
+            tallies = [
+                json.loads(m.body)
+                for m in self.queue_manager.read_messages(self.controller._tallies_queue())
+            ]
+            entities_from_tallies = {(t['entity_id'], t['entity_type']) for t in tallies}
+            entities_from_contributions = {
+                (contributions.entity.entity_id, contributions.entity.entity_type)
+                for cl in contributions_per_bundle for contributions in cl
+            }
+            self.assertSetEqual(entities_from_tallies, entities_from_contributions)
+            events = []
+            for tally in tallies:
```

Comment:
I think it would make sense to split this into two separate test methods, starting the second one here. This isn't necessary, but there is a logical divide here between testing `contribute` and `aggregate`, and the test method is already pretty long.

Reply:
@jessebrennan 
The original plan was to test `contribute` and `aggregate` separately, but @hannes-ucsc  suggested to  just perform the aggregation because we already have the contributions ready to go within this test. 

Perhaps a subtest can be used show distinction from the two different operations?

Reply:
Personally I'm not a fan of using subtest this way, but keep it if you prefer and/or Hannes will decide.

Reply:
Subtests cause the test code to resume after the CM body upon failure in the CM body. So subtests should be used to mark section in the progression. Here I think it makes sense to test contribution and aggregation in the same method. If it were done in a separate method, the method for testing aggregation would duplicate a lot of the code from the method for testing contributions.